Extending LLDB to work on AIX

DhruvSrivastavaX opened the pull request:
We have Implemented the necessary code changes required to run LLDB on AIX. This PR is for discussion as asked in #101657
With this PR, we request your inputs and feedback about our code changes. 

This is also opened for discussion on the discourse platfom:
https://discourse.llvm.org/t/port-lldb-to-ibm-aix/80640

Co-authored by: Felix (Ting Wang) https://github.com/orcguru

DavidSpickett commented on lldb/CMakeLists.txt:
```diff
@@ -38,6 +38,10 @@ endif()
 include(AddLLDB)
 
+if (UNIX AND ${CMAKE_SYSTEM_NAME} MATCHES "AIX")
+  add_definitions("-D__AIX__")
```
This looks like a macro that a compiler might set for you, is that not the case here?

DavidSpickett replied:
Also remember that users may be debugging AIX from a non-AIX host where the compiler would not be setting any macros like that.

(though perhaps this is exactly why you defined this manually, just to get it working on AIX only as the first step)

DavidSpickett commented on lldb/NOTICE.TXT:
```diff
@@ -0,0 +1,7 @@
+This product contains small piece of code to support AIX, taken from netbsd.
+
+  * LICENSE:
+    * lldb/source/Host/common/LICENSE.aix-netbsd.txt (OpenSSL License)
```
We'll need to confirm license compatibility for this.

DavidSpickett replied:
And if there is a way not to pull in this code, we're likely going to prefer that.

DavidSpickett commented on lldb/include/lldb/Host/XML.h:
```diff
@@ -11,6 +11,11 @@
 
+#if defined(__AIX__)
+//FIXME for AIX
+#undef LLDB_ENABLE_LIBXML2
```
In theory if you configure with `-DLLDB_ENABLE_LIBXML2=OFF`, the libxml2 parts will be removed. Was something else being included despite that?

DavidSpickett commented on lldb/include/lldb/Host/aix/Ptrace.h:
```diff
@@ -0,0 +1,62 @@
+#define ARCH_GET_GS 0x1004
+#endif
+#ifndef PTRACE_PEEKMTETAGS
+#define PTRACE_PEEKMTETAGS      (PT_COMMAND_MAX+7)
```
Funny seeing this here, I worked on lldb's MTE support for AArch64. Does AIX support a memory tagging feature as well?

If so I'd be quite interested to see how it fits into the memory tagging support code. I tried to make it generic but when you've only got one implementation, you never know if it's truly flexible or not.

DavidSpickett commented on lldb/include/lldb/Target/RegisterContextUnwind.h:
```diff
@@ -67,6 +67,10 @@ class RegisterContextUnwind : public lldb_private::RegisterContext {
+#ifdef __AIX__
+  bool ReadLR(lldb::addr_t &lr);
+#endif
+
```
It's likely we'd ask you to make this a method on all platforms that is only called on, or implemented for, AIX.

DavidSpickett commented on lldb/source/API/CMakeLists.txt:
```diff
@@ -40,6 +40,113 @@ add_custom_target(lldb-sbapi-dwarf-enums
 set_target_properties(lldb-sbapi-dwarf-enums PROPERTIES FOLDER "LLDB/Tablegenning")
 
+if(CMAKE_SYSTEM_NAME MATCHES "AIX")
+add_lldb_library(liblldb STATIC ${option_framework}
```
I'm sure we can find a way to refactor this, deal with that later, but is there a reason that it needs to be static? Is this a property of libraries on AIX perhaps.

I know we had someone wanting to build a static library elsewhere but I'm not sure if it was this one.

DavidSpickett commented on lldb/source/API/SBBreakpoint.cpp:
```diff
@@ -342,7 +342,7 @@ uint32_t SBBreakpoint::GetIgnoreCount() const {
 }
 
-void SBBreakpoint::SetThreadID(tid_t tid) {
+void SBBreakpoint::SetThreadID(lldb::tid_t tid) {
```
If these lldb:: are fixing compilation errors they would be welcome as a separate patch.

DavidSpickett commented on lldb/source/Core/Module.cpp:
```diff
@@ -1510,6 +1510,18 @@ bool Module::SetLoadAddress(Target &target, lldb::addr_t value,
+    return true;
+  } else {
+    changed = false;
+  }
```
Tip for later PRs, https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

(but don't bother with style changes at this point of course)

DavidSpickett reviewed (commented):
It's worth putting the PR into draft mode just to be extra clear this is a request for early feedback.

This PR will be later split up of course, so very general comments from me. Very impressive effort overall.

DhruvSrivastavaX commented:
> BTW, the main loop changes are something that you should be able to upstream quite easily. It sounds like the main problem is that you don't have a ppoll syscall, but the last round of changes means that ppoll is not actually necessary for its operation (the only reason to use it is the increased precision)

Hmm. That is good idea too. I think we can work on that. Thanks!!

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Removed un-needed changes

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Resolving coredump issue while attach with library calls

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Resolved minor merge conflict
Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge pull request #22 from DhruvSrivastavaX/11_Attach_coredump

Resolving coredump issue while attach with library calls

DhruvSrivastavaX committed:
[AIX][Coredump] AIX Coredump debugging Implementation (#25)

Creates a general framework to be able to debug an AIX generated coredump file.
At this point, we are only supporting 64-bit core files using this general framework.
With this implementation, LLDB can recognise and debug any 64-bit AIX coredump file.

Most of the generic debugging commands work after this:

# bin/lldb --core /home/dhruv/LLDB/tests/core
(lldb) target create --core "/home/dhruv/LLDB/tests/core"
Core file '/home/dhruv/LLDB/tests/core' (powerpc64) was loaded.
(lldb) bt
* thread #1, stop reason = SIGSEGV
  * frame #0: 0x0000000100000940 coretest64`main at core.c:5
    frame #1: 0x00000001000004ac coretest64`__start + 116
(lldb) process status
Process 18446744071562067991 stopped
* thread #1, stop reason = SIGSEGV
    frame #0: 0x0000000100000940 coretest64`main at core.c:5
   2        char *str;
   3        int a = 10;
   4        str = "GfG";
-> 5        *(str+1) = 'n';
   6        return a;
   7     }

And others like memory read, image list, image dump sections, disassembly etc

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Resolving merge conflict(lldb/source/Host/CMakeLists.txt)

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'gh-101657' of https://github.com/Dhruv-Srivastava-IBM/lldb-for-aix into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

ravi-sh committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into llvmgh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Error Handling (#32)

* Error Handling for aix core module

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Resolved merge conflict with file
lldb/tools/lldb-dap/CMakeLists.txt

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
First time attach resolution

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge pull request #36 from DhruvSrivastavaX/12-First-time-attach

First time attach resolution

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Invalid DWARF rangelist

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Fix for stack memory access from core file (#40)

* Fix for stack memory access in core file

DhruvSrivastavaX committed:
Build fail: SBMutex

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge pull request #45 from DhruvSrivastavaX/18-lldb-invalid-range-list-offset

Invalid DWARF rangelist

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Resolving merge conflict "lldb/tools/driver/Driver.cpp"

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Resolved merge conflict in the file
lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Added change for step command issue

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge pull request #53 from DhruvSrivastavaX/43_step_cmd_iss

Added change for step command issue

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Resolved merge conflict with the file:
lldb/tools/lldb-dap/CMakeLists.txt

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Removing netbsd license dependency

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge pull request #59 from DhruvSrivastavaX/removing_dladd

Removing netbsd license dependency

dladdr() code removed.

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Removed netbsd license files

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

HemangGadhavi committed:
Merge pull request #62 from DhruvSrivastavaX/removing_dladd

Removed netbsd license files 
Actual netbsd's code removed through this PR #59

ravindra-shinde2 committed:
Global variables are not accessed correctly

1188732:lldb, print command does not works properly sometimes

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

HemangGadhavi committed:
Get base address for each moudle for image list command

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

DhruvSrivastavaX committed:
Fix for GetName() (#61)

Need to access process/main-thread name from psinfo on aix rather than comm on linux.

(0) root @ unobosdev002: /home/dhruv/dio_fs/LLDB/build-lldb
# bin/lldb ../tests/1
(lldb) target create "../tests/1"
Current executable set to '/home/dhruv/dio_fs/LLDB/tests/1' (powerpc64).
(lldb) b main
Breakpoint 1: where = 1`main + 8 at 1.c:3, address = 0x0000000100000928
(lldb) r
Process 44433678 launched: '/home/dhruv/dio_fs/LLDB/tests/1' (powerpc64)
Process 44433678 stopped
* thread #1, name = '1', stop reason = breakpoint 1.1
    frame #0: 0x0000000100000928 1`main at 1.c:3
   1    int main()
   2    {
-> 3        int x = 10;
   4        x++;
   5        return x;
   6    }
(lldb) thread info
thread #1: tid = 0x2a6010e, 0x0000000100000928 1`main at 1.c:3, name = '1', stop reason = breakpoint 1.1

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Merge pull request #64 from DhruvSrivastavaX/55_image_list

Get base address for each module for image list command

ravindra-shinde2 committed:
Merge pull request #63 from DhruvSrivastavaX/1188732_static_global

Global variables are not accessed correctly, as the data/bss sections corrupted.

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

DhruvSrivastavaX committed:
Stabilizing the branch to fix forced-update issue
- Linked to Recent PRs for aix/Host.cpp, aix/Support.cpp etc

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Re-Design for ObjectFileXCOFF - 64/32 bit (#72)

This is the New design for the ObjectFileXCOFF using LLVM interfaces.

The intent for which I had to implement this is that
we don't need to create new structures and can use LLVM interfaces instead, which makes the code cleaner,
safer and more structured.
And community was not accepting the older design.

As a plus, I have also made sure that it is able to handle 32 bit binaries too along with 64 bit

I have discarded the older design and only kept whats necessary, without changing any existing behaviour.

Most of this code has already been reviewed and merged in the community,
so I am pulling it together again here in our fork branch.

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Conflicts:
lldb/source/Target/RegisterContextUnwind.cpp

DhruvSrivastavaX committed:
Merge typo

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Enabled Code formatting workflow

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge conflict llvm-main Pull (#74)

Merge remote-tracking branch 'upstream/main' into sync-upstream

 Conflicts:
        lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
 ParseSymtab

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

Commit:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp

DhruvSrivastavaX committed:
Build fail due to RemoveBreakpoint change in llvm:main

 Changes to be committed:
	modified:   lldb/source/Plugins/Process/AIX/NativeProcessAIX.cpp

HemangGadhavi committed:
Merge remote-tracking branch 'upstream/main' into merge_0307_

DhruvSrivastavaX committed:
Merge pull request #84 from DhruvSrivastavaX/merge_0307_

Merge remote-tracking branch 'upstream/main' into gh-101657

Conflicts:
lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.cpp
and kept below file as it is not merged from the main
lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp

HemangGadhavi committed:
Merge remote-tracking branch 'upstream/main' into merge_0707

DhruvSrivastavaX committed:
Merge pull request #85 from DhruvSrivastavaX/merge_0707

Merge remote-tracking branch 'upstream/main' into merge_0707

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

ravindra-shinde2 committed:
Fix: Corrected offset to retrieve process name and process ID (#91)

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

ravindra-shinde2 committed:
Defect/corefile issues (#92)

* Resolving redefination Error for the PR #91

Signed-off-by: Ravindra Shinde <Ravindra.Shinde2@ibm.com>

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into merge-04-08

 Conflicts:
	lldb/source/Plugins/Process/CMakeLists.txt

DhruvSrivastavaX committed:
Merge pull request #95 from DhruvSrivastavaX/merge-04-08

Merge 04 08

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
32bit support P1 (#96)

DhruvSrivastavaX committed:
[Core][32 bit] AIX 32 bit coredump support (#97)

DhruvSrivastavaX committed:
[Core][Multi-threaded] Multi-threaded Core file support for 64 and 32 bit (#99)

DhruvSrivastavaX committed:
Merge conflict resolution

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
std:__1:string fix for AIX (#119)

* std:__1:string fix for AIX

Fixes #76

(lldb) n
Process 26214908 stopped
* thread #1, name = '1191468_l', stop reason = step over
    frame #0: 0x00000001000009a4 1191468_l`main at 1191468_l.C:14
   11        std::string str = "Hello";
   12        std::string str_l = "Hello! This is a very very long string that goes beyond the short string limit";
   13        // ...
-> 14        std::cout << "string " << str << std::endl;
   15        std::cout << "string " << str_l << std::endl;
   16    
   17        return 0;
(lldb) p str
(std::__1::string) "Hello"
(lldb) p str_l
(std::__1::string) "Hello! This is a very very long string that goes beyond the short string limit"

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Cmake build update (#120)

For AIX7.2 and
sometimes for AIX7.3 as well (not always)
Build would fail due to unidentified macros in AIX.
Need to link the definitions correctly.

Errors such as these:

In file included from /home/dhruv/LLDB/lldb-for-aix/lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:26:
In file included from /home/dhruv/LLDB/lldb-for-aix/lldb/include/lldb/Breakpoint/Watchpoint.h:18:
In file included from /home/dhruv/LLDB/lldb-for-aix/lldb/include/lldb/Target/Target.h:24:
In file included from /home/dhruv/LLDB/lldb-for-aix/lldb/include/lldb/Core/ModuleList.h:13:
/home/dhruv/LLDB/lldb-for-aix/lldb/include/lldb/Core/ModuleSpec.h:51:27: warning: cast from 'const char *' to 'char *' drops const qualifier [-Wcast-qual]
      char *str = (char *)path_with_object.c_str();
                          ^
In file included from /home/dhruv/LLDB/lldb-for-aix/lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:98:
/usr/include/sys/ldr.h:168:2: error: unknown type name 'uint'; did you mean 'int'?
        __I_FIELDS(__I_INT64, __I_PTR64, __I_FP64, uint ldinfo_flags;)
        ^
/usr/include/sys/ldr.h:137:2: note: expanded from macro '__I_FIELDS'
        uint                    ldinfo_next;        /* offset from current     \
        ^
/usr/include/sys/ldr.h:168:45: error: unknown type name 'uint'; did you mean 'int'?
        __I_FIELDS(__I_INT64, __I_PTR64, __I_FP64, uint ldinfo_flags;)
...
...
...

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
build failure fix post pull (#121)

Changes in DYLD and ProcessAIXCore needed in reference to this PR in the community,
as it caused build failures:

[lldb] Enable locate module callback for all module loading llvm/llvm-project#160199

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Fix for executable error for realtive path lldb (#122)

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Fix for delay while quitting debugging session (#123)

* Fix for delay while quitting debugging session
Closes #47

DhruvSrivastavaX committed:
rerun coredump (#125)

Coredump fix patch for handling new code changes from community on rerun of binary.

(lldb) r
There is a running process, kill it and restart?: [Y/n] y
Process 26739144 exited with status = 9 (0x00000009) 
Assertion failed: (matching_module_list.IsEmpty() || module_spec.GetFileSpec().GetFilename().IsEmpty() || module_spec.GetFileSpec().GetFilename() != matching_module_list.GetModuleAtIndex(0) ->GetFileSpec() .GetFilename()) && "Search by name not found in SharedModuleList's map", file  /home/LLDB/lldb-for-aix/lldb/source/Core/ModuleList.cpp, line 800, void (anonymous namespace)::SharedModuleList::FindModules(const lldb_private::ModuleSpec &, lldb_private::ModuleList &) const()
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: lldb /home/LLDB/tests/2

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
ObjectfileXCOFF update m_data (#126)

Community replacement of Objectfile m_data to m_data_nsp

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
extractor_sp change for BigArchive and AIX core (#127)

Build failure fix due to community design change in Big Archive and ObjectfileAIXCore from DataBufferSP to DataExtractorSP

DhruvSrivastavaX committed:
[Resolve] Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/Process/AIX/NativeProcessAIX.cpp

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/ObjectContainer/Big-Archive/CMakeLists.txt
	lldb/source/Plugins/ObjectContainer/Big-Archive/ObjectContainerBigArchive.cpp
	lldb/source/Plugins/ObjectContainer/Big-Archive/ObjectContainerBigArchive.h
	lldb/source/Plugins/ObjectContainer/CMakeLists.txt
	lldb/source/Plugins/Process/AIX/NativeRegisterContextAIX_ppc64.cpp
	lldb/source/Plugins/Process/AIX/NativeRegisterContextAIX_ppc64.h

DhruvSrivastavaX committed:
disassembly fix for symbol resolution with branch statements. (#129)

Disassembly fix after resolving function symbols:
(lldb) di
1189804`main:
    0x100000898 <+0>:   mflr   0
    0x10000089c <+4>:   std    31, -8(1)
    0x1000008a0 <+8>:   std    0, 16(1)
    0x1000008a4 <+12>:  stdu   1, -176(1)
    0x1000008a8 <+16>:  mr     31, 1
    0x1000008ac <+20>:  li     3, 0
    0x1000008b0 <+24>:  stw    3, 164(31)
->  0x1000008b4 <+28>:  ld     3, 176(2)
    0x1000008b8 <+32>:  bl     0x100000da8               ; symbol stub for: printf
    0x1000008bc <+36>:  ld     2, 40(1)
    0x1000008c0 <+40>:  ld     3, 192(2)
    0x1000008c4 <+44>:  addi   4, 31, 140
    0x1000008c8 <+48>:  bl     0x100000dd0               ; symbol stub for: scanf
    0x1000008cc <+52>:  ld     2, 40(1)
    0x1000008d0 <+56>:  lwz    3, 140(31)
    0x1000008d4 <+60>:  extsw  4, 3
    0x1000008d8 <+64>:  addi   3, 31, 144
    0x1000008dc <+68>:  bl     0x1000009a8               ; modifyUnion at 1189804.c:63
    0x1000008e0 <+72>:  nop    
    0x1000008e4 <+76>:  addi   3, 31, 144
    0x1000008e8 <+80>:  bl     0x100000a88               ; printUnion at 1189804.c:45
    0x1000008ec <+84>:  nop

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/ObjectContainer/Big-Archive/ObjectContainerBigArchive.cpp
	lldb/source/Plugins/ObjectFile/PECOFF/ObjectFilePECOFF.cpp
	lldb/source/Plugins/SymbolFile/DWARF/DWARFFormValue.cpp

DhruvSrivastavaX committed:
Updated files based on community update
	modified:   lldb/source/Plugins/ObjectContainer/Big-Archive/ObjectContainerBigArchive.cpp
	modified:   lldb/source/Plugins/ObjectFile/AIXCore/ObjectFileAIXCore.cpp
	modified:   lldb/source/Plugins/ObjectFile/AIXCore/ObjectFileAIXCore.h
	modified:   lldb/source/Plugins/ObjectFile/PECOFF/ObjectFilePECOFF.cpp

HemangGadhavi committed:
cleaning the AIX native files converting from LE to BE (#135)

This commit updates AIX-specific native files from Little Endian (LE) to Big Endian (BE). This ensures that the AIX implementation uses the correct byte order and maintains consistency across the system.

For Ref: This PR does not make changes to any behavioural functionality.

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into merge-12-02-26

 Conflicts:
	lldb/include/lldb/Host/HostInfoBase.h
	lldb/source/Plugins/Process/Utility/RegisterInfoPOSIX_ppc64.cpp

DhruvSrivastavaX committed:
Merge pull request #142 from DhruvSrivastavaX/merge-12-02-26

     Conflicts:
            lldb/include/lldb/Host/HostInfoBase.h
            lldb/source/Plugins/Process/Utility/RegisterInfoPOSIX_ppc64.cpp

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Align ptrace requests with AIX ptrace IDs and enable FPR/VSX/VMX calls (#139)

* Align ptrace requests with AIX ptrace request IDs and enable FPR/VSX/VMX base calls for live processes

DhruvSrivastavaX committed:
Add version details specific to AIX (#138)

* Add version details specific to AIX

* Bumped up the version to 21.1.3.2

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/Process/CMakeLists.txt

DhruvSrivastavaX committed:
build failure fix (#147)

FreeBSDKernel has been removed from the official main, so removed from cmake

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/ScriptInterpreter/Python/CMakeLists.txt

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/include/lldb/Host/HostInfoBase.h

DhruvSrivastavaX committed:
build fix

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Build failure due to
	modified:   llvm/lib/Transforms/Vectorize/VPlan.h

DhruvSrivastavaX committed:
failure due to tid_t

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Aix 32bit support (#148)

* 32 bit binaries support

- PPC/PPC64 arch detection
- Registers read/write handling

* Instruction emulator support for 32bit

* Addressed review comments

* Added template based get/set SPRS call

DhruvSrivastavaX committed:
[WATCHPOINT] Watchpoint support for AIX (#149)

* Watchpoint support for AIX:
- Added support for Hardware watchpoints on AIX to watch variables

* size protection

* Addressed comments

* optimsation

DhruvSrivastavaX committed:
Bumping up version to 22.1.1.0 (#150)

Version update -> 22.1.1.0

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
LOGV deprecated

HemangGadhavi committed:
Fix coredump on quit after pulling from llvm:main (#157)

DhruvSrivastavaX committed:
Coredump issues with watchpoint rerun and 32-bit coredump loading (#156)

Fixes #155
This commit fixes two separate coredump issues:

Coredump with all platforms (Mac, Linux, AIX etc) with watchpoint rerun on same variable
Coredump with loading 32-bit coredump due to internal design changes.

HemangGadhavi committed:
Fix for stl, vector and static variable access for 32-bit (#158)

* Fix for stl, vector and static variable access for 32-bit

* Review comment

DhruvSrivastavaX committed:
Fix for value size in watch expression (#159)

Fixes #153
Adds size handling for:
watchpoint set expression -s 4 -- 0x0000000110000b18 
and
watchpoint set expression &var

Adds warning for: watchpoint set expression -- 0x0000000110000b18

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

Conflicts:
lldb/source/Plugins/ObjectContainer/Big-Archive/ObjectContainerBigArchive.cpp
lldb/source/Plugins/ObjectFile/Minidump/ObjectFileMinidump.cpp
lldb/source/Plugins/ObjectFile/PDB/ObjectFilePDB.cpp
lldb/source/Plugins/ObjectFile/PECOFF/ObjectFilePECOFF.cpp
lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
lldb/source/Plugins/ObjectFile/AIXCore/ObjectFileAIXCore.cpp
lldb/source/Plugins/ObjectFile/AIXCore/ObjectFileAIXCore.h
lldb/source/Plugins/Process/AIX/NativeProcessAIX.cpp

DhruvSrivastavaX committed:
extractor_sp update for ObjectfileAIXCore (#162)

Fixes #161

(lldb) target create --core "/home/dhruv/tests_DBX/core"
Core file '/home/dhruv/tests_DBX/core' (powerpc64) was loaded.
(lldb) bt
* thread #1, name = 'TS021241538', stop reason = SIGSEGV
  * frame #0: 0x0000000100000948 TS021241538`level3(value=3) at TS021241538.cpp:11
    frame #1: 0x00000001000009dc TS021241538`level2(value=2) at TS021241538.cpp:20
    frame #2: 0x0000000100000a5c TS021241538`level1(value=1) at TS021241538.cpp:26
    frame #3: 0x0000000100000ad0 TS021241538`main at TS021241538.cpp:32
    frame #4: 0x00000001000004ac TS021241538`__start + 116
(lldb) q

DhruvSrivastavaX committed:
Updating version to 22.1.2.0

HemangGadhavi committed:
Merge remote-tracking branch 'upstream/main' into merge_08

DhruvSrivastavaX committed:
Merge pull request #164 from DhruvSrivastavaX/merge_08

Merge remote-tracking branch 'upstream/main' into merge_08

PR upstreaming Conflicts:
source/Plugins/Process/AIX/NativeProcessAIX.cpp
source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp
source/Host/aix/Host.cpp

HemangGadhavi committed:
Resolved build failure with latest merge

DhruvSrivastavaX committed:
Merge pull request #165 from DhruvSrivastavaX/build_f

Resolved build failure with latest merge

AsCString(nullptr)

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Build fix

HemangGadhavi committed:
[Regression] Fix for path resolution when attaching to a process due to 32-bit design change (#167)

* Fix for path resolution when attaching to a process due to 32-bit design change

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/Process/gdb-remote/CMakeLists.txt
	lldb/source/Target/RegisterContextUnwind.cpp

DhruvSrivastavaX committed:
Design change support for coredump (#170)

*32-bit design change integration had missed coredump handling, so it was encountering issues:
Updated the design

Co-authored-by: Hemang Gadhavi <hemang.gadhavi@ibm.com>

HemangGadhavi committed:
Merge remote-tracking branch 'upstream/main' into merge_30_04

DhruvSrivastavaX committed:
Merge pull request #173 from DhruvSrivastavaX/merge_30_04

Resolved merge conflits :

source/Plugins/Process/AIX/NativeProcessAIX.cpp
source/Plugins/Process/AIX/NativeRegisterContextAIX.cpp
source/Plugins/Process/AIX/NativeRegisterContextAIX.h
source/Plugins/Process/AIX/NativeRegisterContextAIX_ppc64.cpp
source/Plugins/Process/AIX/NativeRegisterContextAIX_ppc64.h
source/Plugins/Process/AIX/NativeThreadAIX.cpp
source/Plugins/Process/AIX/NativeThreadAIX.h
source/Plugins/Process/Utility/RegisterInfos_ppc64.h
source/ValueObject/ValueObject.cpp

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/ScriptInterpreter/Python/CMakeLists.txt
	lldb/tools/lldb-dap/CMakeLists.txt

HemangGadhavi committed:
Merge remote-tracking branch 'upstream/main' into merge_19_05

DhruvSrivastavaX committed:
Merge pull request #177 from DhruvSrivastavaX/merge_19_05

Merge remote-tracking branch 'upstream/main' into merge_19_05

Conflicts:
lldb/tools/driver/Driver.cpp

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Commands/CommandObjectWatchpoint.cpp

DhruvSrivastavaX committed:
Changes to be committed:
	modified:   lldb/source/Commands/CommandObjectWatchpoint.cpp
	modified:   lldb/source/Host/common/File.cpp

DhruvSrivastavaX committed:
[Multi-threaded] Fixing the PID vs TID mixup in Native Layer (#178)

* [Multi-threaded] Fixing the PID vs TID mixup in Native Layer

* Updated DiscoverThreads with llvm APIs

* Replaced PT_CONTINUE with PTT_CONTINUE

HemangGadhavi committed:
Add fix for next issue with latest compiler (#179)

LLDB was not placing the breakpoint on the next instruction when executing the next command. 
This issue is only reproducible with the latest compiler, as recent compiler changes have modified the generated instruction sequence.
Added the macro to validate the lower address range.

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Version updated to 22.1.6.0

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Add fix for global var access issue with latest compiler (#182)

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Regression-Fix for core stack not displaying for 32&64 for 22.1.2.4 (#187)

Regression fix for PRs:

Add fix for global var access issue with latest compiler #182 and  #183

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Build failure fix 7-july-26

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/tools/driver/CMakeLists.txt

HemangGadhavi committed:
Fix for Redundant entries of libraries on re-run (#185)

* Fix for Redundant entries of libraries on re-run
* adressed review comment

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

HemangGadhavi committed:
Add fix for SAP attach iss (#194)

Splits the process and its parameter

DhruvSrivastavaX committed:
Watchpoint assertion defect (#197)

Fixes #184

DhruvSrivastavaX committed:
Resume and Step logic Modification as per AIX design  v 22.1.7.0 (#202)

* Resume and Step logic Modification as per AIX design
And Cleanup of dead code/comments

* Version update: 22.1.7.0

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657

DhruvSrivastavaX committed:
Build failure fix after merge
 Changes to be committed:
	modified:   lldb/source/API/CMakeLists.txt
	modified:   lldb/source/Plugins/ObjectFile/XCOFF/ObjectFileXCOFF.cpp

DhruvSrivastavaX committed:
Symbol Resolution fixes: (#205)

Removes un-needed label symbols and hidden entries from the symbol
table saved in lldb to avoid symbol namespace pollution.
Additionally, adds more accessibility to aix core to be able to read
multiple memory regions (data, sdata)

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/source/Plugins/ObjectContainer/Big-Archive/ObjectContainerBigArchive.h

DhruvSrivastavaX committed:
Build failure fix

DhruvSrivastavaX committed:
Vtable update for C++ Symbols (#210)

Updating gh to 22.1.7.3 changes:

C++ dynamic type resolution #209

DhruvSrivastavaX committed:
DWARF Forms fix for AIX (#211)

DW_FORM_strp, DW_FORM_line_strp, DW_FORM_sec_offset

DhruvSrivastavaX committed:
Typo correction

DhruvSrivastavaX committed:
Attach path fix (execv) (#214)

* Attach path fix (execv)

* Removed additional /

DhruvSrivastavaX committed:
Enhanced core loader handling (#215)

* Loader fix for core

* Added loader_size guard

DhruvSrivastavaX committed:
Merge remote-tracking branch 'upstream/main' into gh-101657

 Conflicts:
	lldb/include/lldb/Host/HostInfoBase.h
	lldb/source/Plugins/Process/AIX/NativeProcessAIX.cpp
	lldb/source/Plugins/Process/AIX/NativeProcessAIX.h

DhruvSrivastavaX committed:
Merge branch 'llvm:main' into gh-101657
